File: r/computeserver.R
require("httpuv")
require("jsonlite")
require("curl")
require("purrr")

serverState = list()
serverState$isRunning = TRUE

toMime = list()
toMime$json = "application/json;charset=utf-8"
toMime$jsonl = "application/json;charset=utf-8"
toMime$csv = "text/csv;charset=utf-8"
toMime$txt = "text/plain;charset=utf-8"
toMime$bin = "application/octet-stream"
dataFormatToMime = function(dataFormat)
{
  mime = toMime[dataFormat][[1]]
  if (is.null(mime)) mime = ""
  mime
}

# data caching
cache = list()
cache$map = list()
cache$exists = function(uri)
{
  !is.null(cache$map[[uri]])
}
cache$add = function(uri, format, value)
{
  cache$map[[uri]] <<- list(format=format, value=value)
}
cache$get = function(uri)
{
  cache$map[[uri]]
}

# data conversion
inputFromText = function(text, format)
{
  x = NULL
  
  if (format == "csv")
  {
    x = read.csv(textConnection(text), header=T)
  }
  else if (format == "json")
  {
    x = fromJSON(text)
  }
  else if (format == "jsonl")
  {
    x = stream_in(textConnection(text))
  }
  else if (format == "txt")
  {
    x = readLines(textConnection(text))
  }
  else if (format == "bin")
  {
    x = base64_dec(text)
  }
  x
}
inputFromUrl = function(url, format)
{
  x = NULL
  
  if (format == "csv")
  {
    x = read.csv(url, header=T)
  }
  else if (format == "json")
  {
    x = fromJSON(url)
  }
  else if (format == "jsonl")
  {
    x = stream_in(url(url))
  }
  else if (format == "txt")
  {
    x = readLines(url(url))
  }
  else if (format == "bin")
  {
    x = readBin(url, "raw", n=20000000)
  }
  x
}
output = function(dataObject, format)
{
  x = ""
  if (format == "csv")
  {
    con = textConnection("tc1", open="w")
    write.csv(dataObject, con, row.names=F)
    x = paste(textConnectionValue(con), collapse="\n")
    close(con)
  }
  else if (format == "json")
  {
    x = toJSON(dataObject, auto_unbox=T) # auto_unbox: don't put simple data types into json array
  }
  else if (format == "jsonl")
  {
    con = textConnection("tc1", open="w")
    stream_out(dataObject, con)
    x = paste(textConnectionValue(con), collapse="\n")
    close(con)
  }
  else if (format == "txt")
  {
    con = textConnection("tc1", open="w")
    writeLines(dataObject, con)
    x = paste(textConnectionValue(con), collapse="\n")
    close(con)
  }
  else if (format == "bin")
  {
    x = base64_enc(dataObject)
  }
  x
}

koralCompute = function(uri, ds, cache)
{
  get(ds$func) # throws an error when function not found
  
  # load input
  params = list()
  for (name in names(ds$params))
  {
    p = ds$params[[name]]
    value = NULL
    if (!is.null(p$uri) && cache$exists(p$uri))
    {
      value = cache$get(p$uri)$value
    }
    else if (!is.null(p$val))
    {
      if (p$type == "json")
      {
        value = p$val # already converted to json object representation
      }
      else
      {
        valueStr = toJSON(p$val) # back to string representation
        value = inputFromText(valueStr, p$type);
      }
    }
    else if (!is.null(p$url))
    {
      value = inputFromUrl(p$url, p$type)
    }
    if (!is.null(value))
    {
      params[name] = value
    }
  }
  
  # execute
  result = invoke(ds$func, params)
  
  if (!is.null(result)) cache$add(uri, ds$type, result)
  result
}

addCode = function(expressionText)
{
  eval(expr=parse(text=expressionText), envir=globalenv())
}

http_badRequest = list(status = 400L)
http_NotFound = list(status = 404L)
http_okEmpty = list(status = 200L, headers = list('Content-Type' = 'text/html'), body = "")

parseQuery = function(query) 
{
  if (is.null(query) || nchar(query) == 0) return(list())
  if (substring(query, 1, 1) == '?') query = substring(query, 2)
  if (nchar(query) == 0) return (list())
  
  pairs = strsplit(strsplit(query, "&")[[1]], "=")
  params = list()
  for (i in 1:length(pairs))
  {
    kv = pairs[[i]]
    if (length(kv) == 1) next
    k = kv[1]
    v = kv[2]
    if (is.null(k) || nchar(k) == 0 || is.null(v) || nchar(v) == 0) next
    params[URLdecode(k)] = URLdecode(v)
  }
  params
}

app = list(
  call = function(req) {
    verb = req$REQUEST_METHOD
    uri = substring(req$PATH_INFO, 2)
    params = parseQuery(req$QUERY_STRING)
    
    if (verb == "GET")
    {
      if (params$action == "shutDown")
      {
        serverState$isRunning <<- FALSE
        return (http_okEmpty)
      }
      
      if (cache$exists(uri))
      {
        d = cache$get(uri)
        s = output(d$value, d$format)
        return (list(status = 200L, 
                headers=list("Content-Type"= dataFormatToMime(d$format)), 
                body = s))
      }
      
      return (http_NotFound)
    }
    else if (verb == "POST" && params$action == "compute" && nchar(uri) > 0)
    {
      ds = fromJSON(req$rook.input$read_lines())
      result = koralCompute(uri, ds, cache)
      resultStr = ""
      if (is.null(ds$nostore) || !ds$nostore) resultStr = output(result, ds$type)
      
      return (list(status = 200L, 
                   headers=list("Content-Type"= dataFormatToMime(ds$type)), 
                   body = resultStr))
    }
    else if (verb == "POST" && params$action == "addCode") # for unit tests
    {
      code = req$rook.input$read_lines()
      addCode(code)
      return (http_okEmpty)
    }

    http_badRequest
  }
)

print(.libPaths())
args = commandArgs(trailingOnly=T)
if (length(args) != 1)
{
  print("Specify port.")
} else
{
  port = as.numeric(args)
  interruptIntervalMs = ifelse(interactive(), 100, 1000)
  
  server <- startServer(host="127.0.0.1", port=port, app=app)
  serverState$isRunning = TRUE
  while (serverState$isRunning) 
  {
    service(interruptIntervalMs)
    Sys.sleep(0.001)
  }
  stopServer(server)
  print("Server stopped.")
}

File: python/computeserver.py
#! python3
import sys
import time
import io
import base64
import threading
import http.server
import urllib
import json
import pandas as pd

print("python compute server.")


toMime = { 
	"json": "application/json;charset=utf-8",
	"jsonl": "application/json;charset=utf-8",
	"csv": "text/csv;charset=utf-8",
	"txt": "text/plain;charset=utf-8",
	"bin": "application/octet-stream" 
}
def dataFormatToMime(dataFormat):
	toMime.get(dataFormat, default="text/plain;charset=utf-8")

server = None
cache = {}

def shutDown():
	time.sleep(2)
	print("shutDown server...")
	if server is not None: 
		server.shutdown()

def inputFromStr(text, format):
    dataObject = None
    if format == "csv":
        dataObject = read_csv(io.StringIO(text))
    elif format == "json":
        dataObject = json.loads(text)
    elif format == "jsonl":
        dataObject = []
        for x in text.split("\n"):
            dataObject.append(json.loads(x))
    elif format == "txt":
        dataObject = text.split("\n")
    elif format == "bin":
        dataObject = base64.b64decode(text)
    
    return dataObject   

def outputToStr(dataObject, format):
    text = ""
    if format == "csv":
        text = object.to_csv(index=False)  # StringIO.getvalue()
    elif format == "json":
        # pandas dataframe
        if callable(getattr(dataObject, "to_json", None)):
            text = object.to_json()
        else:
            # any object
            text = json.dumps(dataObject)
    elif format == "jsonl":
        # pandas dataframe
        if callable(getattr(dataObject, "to_json", None)):
            text = object.to_json(lines=True)
        else:
            # list or generator
            lines = []
            for x in dataObject:
                lines.append(json.dumps(x))
            text = "\n".join(lines)
    elif format == "txt":
        lines = []
        for x in dataObject:
            lines.append(str(x))
        text = "\n".join(lines)
    elif format == "bin":
    	text = base64.b64encode(dataObject)
    
    return text

def koralCompute(uri, ds, cache):
	args = {}
	for name in ds["params"].keys():
		p = ds["params"][name]
		value = None
		if "uri" in p.keys() and p["uri"] in cache.keys():
			value = cache[p["uri"]][0]
		elif "val" in p.keys():
			if p["type"] == "json":
				value = p["val"]
			else:
				value = inputFromStr(p["val"], p["type"])
		elif "url" in p.keys():
			valueStr = urllib.urlopen(p["url"]).read()
			value = inputFromStr(valueStr, p["type"])

		if value is not None:
			args[name] = value
	
	return globals()[ds["func"]](**args)

def addCode(codeText):
	exec(codeText)

class ComputeServer(http.server.BaseHTTPRequestHandler):
	def getUri(self):
		req = urllib.parse.urlparse(self.path)
		return req.path[1:]

	def getAction(self):
		req = urllib.parse.urlparse(self.path)
		q = urllib.parse.parse_qs(req.query)
		action = ""
		if "action" in q.keys():
			l = q["action"]
			if (len(l) == 1):
				action = l[0]
		return action	

	def do_GET(self):
		uri = self.getUri()
		action = self.getAction()
		print("GET: " + self.path + " " + action)

		if action == "shutDown":
			threading.Thread(target=shutDown).start()
			self.send_response(200)
			return

		if len(uri) < 1:
			self.send_response(400)
			return

		if uri in cache.keys():
			d = cache[uri]
			s = outputToStr(d[0], d[1])
			self.send_response(200)
			self.send_header('Content-type', dataFormatToMime(d[1]))
			self.end_headers()
			self.wfile.write(s.encode("utf-8"))
			return

		self.send_response(404)

	def do_POST(self):
		uri = self.getUri()
		action = self.getAction()

		if action == "compute":
			ds = json.loads(self.rfile.read().decode())

			if len(uri) < 1:
				self.send_response(400)
				return

			result = koralCompute(uri, ds, cache)
			self.send_response(200)

			if not ds.get("nostore", default=false):
				self.send_header('Content-type', dataFormatToMime(ds.type))
				self.end_headers()
				self.wfile.write(outputToStr(result, ds.type).encode("utf-8"))
			return

		if action == "addCode":
			print("addCode")
			addCode(self.rfile.read().decode())
			self.send_response(200)
			return

		self.send_response(400)

if len(sys.argv) != 2:
	sys.exit("USAGE: port")

port = int(sys.argv[1])
server = http.server.HTTPServer(("", port), ComputeServer)

#try:
#    server.serve_forever()
#except KeyboardInterrupt:
#    pass
#server.server_close()
#print(time.asctime(), "Server stopped.")

threading.Thread(target=server.serve_forever).start()
print(time.asctime(), "Server started.")